src/Zigzag_persistence/include/gudhi/zigzag_persistence.h

```diff
@@ -0,0 +1,458 @@
+  index insert_face(const BoundaryRange& boundary, dimension_type dimension) {
+    ++numArrow_;
+    _process_forward_arrow(boundary, dimension);
+    return numArrow_;
+  }
+
+  /**
+   * @brief Updates the zigzag persistence diagram after the removal of the given face.
+   *
+   * @param arrowNumber Arrow number of when the face to remove was inserted for the last time.
+   * @param dimension Dimension of the face to remove.
+   * @return Number of the operation.
+   */
+  index remove_face(index arrowNumber, dimension_type dimension) {
+    ++numArrow_;
+    _process_backward_arrow(arrowNumber, dimension);
+    return numArrow_;
+  }
+
+  /**
+   * @brief To use when a face is neither inserted nor removed, but the filtration moves along the identity operator
+   * on homology level. Useful to keep the birth/death indices aligned when insertions/removals are purposely skipped
+   * to avoid useless computation. Increases the arrow number by one.
+   * @return Number of the operation.
+   */
+  index apply_identity() { return ++numArrow_; }
+
+  /**
+   * @brief Outputs through the given callback method all birth indices which are currently not paired with 
+   * a death index.
+   * 
+   * @tparam F Type of the callback method. Takes two arguments: the dimension of the cycle and the birth index
+   * of the cycle.
+   * @param stream_infinite_interval Method processing the unpaired birth indices.
+   */
+  template <typename F>
+  void get_current_infinite_intervals(F&& stream_infinite_interval) {
+    for (auto& p : births_) {
+      if constexpr (erase_birth_history) {
+        auto& col = matrix_.get_column(p.first);
+        stream_infinite_interval(col.get_dimension(), p.second);
+      } else {
+        try {
+          auto& col = matrix_.get_column(p.first);
+          if (!col.is_paired()) {
+            stream_infinite_interval(col.get_dimension(), p.second);
+          }
+        } catch (const std::out_of_range&) {
+          continue;
+        }
+      }
+    }
+  }
+
+ private:
+  /**
+   * @brief Express the boundary cycle of the new face as a sum of cycles in a matrix.
+   * If some cycles are not boundary cycles, i.e., columns with F-index
+   * in the matrix, it applies a surjective diamond to the zigzag module.
+   *
+   * @param boundary Boundary of the inserted face.
+   * @param dim Dimension of the inserted face.
+   */
+  template <class BoundaryRange>
+  void _process_forward_arrow(const BoundaryRange& boundary, dimension_type dim) {
+    std::vector<matrix_index> chainsInF = matrix_.insert_boundary(numArrow_, boundary);
+
+    if (!chainsInF.empty()) {
+      _apply_surjective_reflection_diamond(dim, chainsInF);
+    } else {
+      birthOrdering_.add_birth_forward(numArrow_);
+      if constexpr (erase_birth_history) {
+        births_.emplace_hint(births_.end(), matrix_.get_column_with_pivot(numArrow_), numArrow_);
+      } else {
+        births_[matrix_.get_column_with_pivot(numArrow_)] = numArrow_;
+      }
```

Comment:
I am trying to understand why the 2 cases differ. If we don't have an entry for `matrix_.get_column_with_pivot(numArrow_)` in the map, they both do the same. Now, can we already have an entry? Only when `!erase_birth_history)` I think, so `births_[matrix_.get_column_with_pivot(numArrow_)] = numArrow_;` should work in both cases.
(trying to kill the `if constexpr` 1 at a time :gun: )

Reply:
The two cases don't differ. They did before, but it was a bug. Honestly, I just added the `if constexpr (erase_birth_history)` to be able to benchmark easily the difference with and without, so I went probably a bit fast on this.